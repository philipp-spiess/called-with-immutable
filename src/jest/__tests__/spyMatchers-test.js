/**
 * Copyright (c) 2014-present, Facebook, Inc. All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * @emails oncall+jsinfra
 */

import { overwriteDefaults } from "../../";
overwriteDefaults();

[
  ["toHaveBeenCalled", "jasmine.createSpy"],
  ["toHaveBeenCalled", "jest.fn"],
  ["toBeCalled", "jasmine.createSpy"],
  ["toBeCalled", "jest.fn"]
].forEach(([called, mockName]) => {
  test(`${called} works with ${mockName}`, () => {
    const fn = mockName === "jest.fn" ? jest.fn() : jasmine.createSpy("fn");

    expect(fn).not[called]();
    expect(() => expect(fn)[called]()).toThrowErrorMatchingSnapshot();

    fn();
    expect(fn)[called]();
    expect(() => expect(fn).not[called]()).toThrowErrorMatchingSnapshot();

    expect(() => expect(fn)[called](555)).toThrowErrorMatchingSnapshot();
  });
});

describe("toHaveBeenCalledTimes", () => {
  it("accepts only numbers", () => {
    const fn = jasmine.createSpy("fn");
    fn();
    expect(fn).toHaveBeenCalledTimes(1);

    [{}, [], true, "a", new Map(), () => {}].forEach(value => {
      expect(() =>
        expect(fn).toHaveBeenCalledTimes(value)
      ).toThrowErrorMatchingSnapshot();
    });
  });

  it("verifies that actual is a Spy", () => {
    const fn = function fn() {};

    expect(() =>
      expect(fn).toHaveBeenCalledTimes(2)
    ).toThrowErrorMatchingSnapshot();
  });

  it("works both for Mock functions and Spies", () => {
    [jasmine.createSpy("fn"), jest.fn()].forEach(fn => {
      fn();
      fn();
      expect(fn).toHaveBeenCalledTimes(2);
    });
  });

  it("passes if function called equal to expected times", () => {
    const fn = jasmine.createSpy("fn");
    fn();
    fn();

    expect(fn).toHaveBeenCalledTimes(2);

    expect(() =>
      expect(fn).not.toHaveBeenCalledTimes(2)
    ).toThrowErrorMatchingSnapshot();
  });

  it("fails if function called more than expected times", () => {
    const fn = jasmine.createSpy("fn");
    fn();
    fn();
    fn();

    expect(fn).toHaveBeenCalledTimes(3);
    expect(fn).not.toHaveBeenCalledTimes(2);

    expect(() =>
      expect(fn).toHaveBeenCalledTimes(2)
    ).toThrowErrorMatchingSnapshot();
  });

  it("fails if function called less than expected times", () => {
    const fn = jasmine.createSpy("fn");
    fn();

    expect(fn).toHaveBeenCalledTimes(1);
    expect(fn).not.toHaveBeenCalledTimes(2);

    expect(() =>
      expect(fn).toHaveBeenCalledTimes(2)
    ).toThrowErrorMatchingSnapshot();
  });
});

[
  "lastCalledWith",
  "toBeCalled",
  "toBeCalledWith",
  "toHaveBeenCalled",
  "toHaveBeenCalledWith",
  "toHaveBeenLastCalledWith"
].forEach(calledWith => {
  test(`${calledWith} works only on spies or jest.fn`, () => {
    const fn = function fn() {};

    expect(() => expect(fn)[calledWith]()).toThrowErrorMatchingSnapshot();
  });
});

[
  ["lastCalledWith", "jest.fn"],
  ["toBeCalledWith", "jest.fn"],
  ["toHaveBeenCalledWith", "jasmine.createSpy"],
  ["toHaveBeenCalledWith", "jest.fn"],
  ["toHaveBeenLastCalledWith", "jasmine.createSpy"],
  ["toHaveBeenLastCalledWith", "jest.fn"]
].forEach(([calledWith, mockName]) => {
  const getFunction = () => {
    return mockName === "jest.fn" ? jest.fn() : jasmine.createSpy("fn");
  };

  test(`${calledWith} works with ${mockName} when not called`, () => {
    const fn = getFunction();
    expect(fn).not[calledWith]("foo", "bar");

    expect(() =>
      expect(fn)[calledWith]("foo", "bar")
    ).toThrowErrorMatchingSnapshot();
  });

  test(`${calledWith} works with ${mockName} and no arguments`, () => {
    const fn = getFunction();
    fn();
    expect(fn)[calledWith]();
  });

  test(`${calledWith} works with ${mockName} and arguments that don't match`, () => {
    const fn = getFunction();
    fn("foo", "bar1");

    expect(fn).not[calledWith]("foo", "bar");

    expect(() =>
      expect(fn)[calledWith]("foo", "bar")
    ).toThrowErrorMatchingSnapshot();
  });

  test(`${calledWith} works with ${mockName} and arguments that match`, () => {
    const fn = getFunction();
    fn("foo", "bar");

    expect(fn)[calledWith]("foo", "bar");

    expect(() =>
      expect(fn).not[calledWith]("foo", "bar")
    ).toThrowErrorMatchingSnapshot();
  });

  test(`${calledWith} works with ${mockName} and many arguments that don't match`, () => {
    const fn = getFunction();
    fn("foo", "bar1");
    fn("foo", "bar2");
    fn("foo", "bar3");

    expect(fn).not[calledWith]("foo", "bar");

    expect(() =>
      expect(fn)[calledWith]("foo", "bar")
    ).toThrowErrorMatchingSnapshot();
  });

  test(`${calledWith} works with ${mockName} and many arguments`, () => {
    const fn = getFunction();
    fn("foo1", "bar");
    fn("foo", "bar1");
    fn("foo", "bar");

    expect(fn)[calledWith]("foo", "bar");

    expect(() =>
      expect(fn).not[calledWith]("foo", "bar")
    ).toThrowErrorMatchingSnapshot();
  });
});
