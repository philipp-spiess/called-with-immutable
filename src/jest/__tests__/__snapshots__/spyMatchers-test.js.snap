// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lastCalledWith works only on spies or jest.fn 1`] = `
"<dim>expect(<red>jest.fn()</><dim>)[.not].lastCalledWith(<dim>)

<red>jest.fn()</> value must be a mock function or spy.
Received:
  function: <red>[Function fn]</>"
`;

exports[`lastCalledWith works with jest.fn and arguments that don't match 1`] = `
"<dim>expect(<red>jest.fn()</><dim>).lastCalledWith(<green>expected</><dim>)

Expected mock function to have been last called with:
  <green>[\\"foo\\", \\"bar\\"]</>
But it was last called with:
  <red>[\\"foo\\", \\"bar1\\"]</>"
`;

exports[`lastCalledWith works with jest.fn and arguments that match 1`] = `
"<dim>expect(<red>jest.fn()</><dim>).not.lastCalledWith(<green>expected</><dim>)

Expected mock function to not have been last called with:
  <green>[\\"foo\\", \\"bar\\"]</>"
`;

exports[`lastCalledWith works with jest.fn and many arguments 1`] = `
"<dim>expect(<red>jest.fn()</><dim>).not.lastCalledWith(<green>expected</><dim>)

Expected mock function to not have been last called with:
  <green>[\\"foo\\", \\"bar\\"]</>"
`;

exports[`lastCalledWith works with jest.fn and many arguments that don't match 1`] = `
"<dim>expect(<red>jest.fn()</><dim>).lastCalledWith(<green>expected</><dim>)

Expected mock function to have been last called with:
  <green>[\\"foo\\", \\"bar\\"]</>
But it was last called with:
  <red>[\\"foo\\", \\"bar3\\"]</>
and <red>two more calls</>."
`;

exports[`lastCalledWith works with jest.fn when not called 1`] = `
"<dim>expect(<red>jest.fn()</><dim>).lastCalledWith(<green>expected</><dim>)

Expected mock function to have been last called with:
  <green>[\\"foo\\", \\"bar\\"]</>
But it was <red>not called</>."
`;

exports[`toBeCalled works only on spies or jest.fn 1`] = `
"<dim>expect(<red>jest.fn()</><dim>)[.not].toBeCalled(<dim>)

<red>jest.fn()</> value must be a mock function or spy.
Received:
  function: <red>[Function fn]</>"
`;

exports[`toBeCalled works with jasmine.createSpy 1`] = `
"<dim>expect(<red>spy</><dim>).toBeCalled(<dim>)

Expected spy to have been called."
`;

exports[`toBeCalled works with jasmine.createSpy 2`] = `
"<dim>expect(<red>spy</><dim>).not.toBeCalled(<dim>)

Expected spy not to be called but it was called with:
  <red>Array []</>"
`;

exports[`toBeCalled works with jasmine.createSpy 3`] = `
"<dim>expect(<red>received</><dim>)[.not].toBeCalled(<dim>)

Matcher does not accept any arguments.
Got:
  number: <green>555</>"
`;

exports[`toBeCalled works with jest.fn 1`] = `
"<dim>expect(<red>jest.fn()</><dim>).toBeCalled(<dim>)

Expected mock function to have been called."
`;

exports[`toBeCalled works with jest.fn 2`] = `
"<dim>expect(<red>jest.fn()</><dim>).not.toBeCalled(<dim>)

Expected mock function not to be called but it was called with:
  <red>Array []</>"
`;

exports[`toBeCalled works with jest.fn 3`] = `
"<dim>expect(<red>received</><dim>)[.not].toBeCalled(<dim>)

Matcher does not accept any arguments.
Got:
  number: <green>555</>"
`;

exports[`toBeCalledWith works only on spies or jest.fn 1`] = `
"<dim>expect(<red>jest.fn()</><dim>)[.not].toBeCalledWith(<dim>)

<red>jest.fn()</> value must be a mock function or spy.
Received:
  function: <red>[Function fn]</>"
`;

exports[`toBeCalledWith works with jest.fn and arguments that don't match 1`] = `
"<dim>expect(<red>jest.fn()</><dim>).toBeCalledWith(<green>expected</><dim>)

Expected mock function to have been called with:
  <green>[\\"foo\\", \\"bar\\"]</>
But it was called with:
  <red>[\\"foo\\", \\"bar1\\"]</>"
`;

exports[`toBeCalledWith works with jest.fn and arguments that match 1`] = `
"<dim>expect(<red>jest.fn()</><dim>).not.toBeCalledWith(<green>expected</><dim>)

Expected mock function not to have been called with:
  <green>[\\"foo\\", \\"bar\\"]</>"
`;

exports[`toBeCalledWith works with jest.fn and many arguments 1`] = `
"<dim>expect(<red>jest.fn()</><dim>).not.toBeCalledWith(<green>expected</><dim>)

Expected mock function not to have been called with:
  <green>[\\"foo\\", \\"bar\\"]</>"
`;

exports[`toBeCalledWith works with jest.fn and many arguments that don't match 1`] = `
"<dim>expect(<red>jest.fn()</><dim>).toBeCalledWith(<green>expected</><dim>)

Expected mock function to have been called with:
  <green>[\\"foo\\", \\"bar\\"]</>
But it was called with:
  <red>[\\"foo\\", \\"bar3\\"]</>, <red>[\\"foo\\", \\"bar2\\"]</>, <red>[\\"foo\\", \\"bar1\\"]</>"
`;

exports[`toBeCalledWith works with jest.fn when not called 1`] = `
"<dim>expect(<red>jest.fn()</><dim>).toBeCalledWith(<green>expected</><dim>)

Expected mock function to have been called with:
  <green>[\\"foo\\", \\"bar\\"]</>
But it was <red>not called</>."
`;

exports[`toHaveBeenCalled works only on spies or jest.fn 1`] = `
"<dim>expect(<red>jest.fn()</><dim>)[.not].toHaveBeenCalled(<dim>)

<red>jest.fn()</> value must be a mock function or spy.
Received:
  function: <red>[Function fn]</>"
`;

exports[`toHaveBeenCalled works with jasmine.createSpy 1`] = `
"<dim>expect(<red>spy</><dim>).toHaveBeenCalled(<dim>)

Expected spy to have been called."
`;

exports[`toHaveBeenCalled works with jasmine.createSpy 2`] = `
"<dim>expect(<red>spy</><dim>).not.toHaveBeenCalled(<dim>)

Expected spy not to be called but it was called with:
  <red>Array []</>"
`;

exports[`toHaveBeenCalled works with jasmine.createSpy 3`] = `
"<dim>expect(<red>received</><dim>)[.not].toHaveBeenCalled(<dim>)

Matcher does not accept any arguments.
Got:
  number: <green>555</>"
`;

exports[`toHaveBeenCalled works with jest.fn 1`] = `
"<dim>expect(<red>jest.fn()</><dim>).toHaveBeenCalled(<dim>)

Expected mock function to have been called."
`;

exports[`toHaveBeenCalled works with jest.fn 2`] = `
"<dim>expect(<red>jest.fn()</><dim>).not.toHaveBeenCalled(<dim>)

Expected mock function not to be called but it was called with:
  <red>Array []</>"
`;

exports[`toHaveBeenCalled works with jest.fn 3`] = `
"<dim>expect(<red>received</><dim>)[.not].toHaveBeenCalled(<dim>)

Matcher does not accept any arguments.
Got:
  number: <green>555</>"
`;

exports[`toHaveBeenCalledTimes accepts only numbers 1`] = `
"<dim>expect(<red>received</><dim>)[.not].toHaveBeenCalledTimes(<green>expected</><dim>)

Expected value must be a number.
Got:
  object: <green>{}</>"
`;

exports[`toHaveBeenCalledTimes accepts only numbers 2`] = `
"<dim>expect(<red>received</><dim>)[.not].toHaveBeenCalledTimes(<green>expected</><dim>)

Expected value must be a number.
Got:
  array: <green>Array []</>"
`;

exports[`toHaveBeenCalledTimes accepts only numbers 3`] = `
"<dim>expect(<red>received</><dim>)[.not].toHaveBeenCalledTimes(<green>expected</><dim>)

Expected value must be a number.
Got:
  boolean: <green>true</>"
`;

exports[`toHaveBeenCalledTimes accepts only numbers 4`] = `
"<dim>expect(<red>received</><dim>)[.not].toHaveBeenCalledTimes(<green>expected</><dim>)

Expected value must be a number.
Got:
  string: <green>\\"a\\"</>"
`;

exports[`toHaveBeenCalledTimes accepts only numbers 5`] = `
"<dim>expect(<red>received</><dim>)[.not].toHaveBeenCalledTimes(<green>expected</><dim>)

Expected value must be a number.
Got:
  map: <green>Map {}</>"
`;

exports[`toHaveBeenCalledTimes accepts only numbers 6`] = `
"<dim>expect(<red>received</><dim>)[.not].toHaveBeenCalledTimes(<green>expected</><dim>)

Expected value must be a number.
Got:
  function: <green>[Function anonymous]</>"
`;

exports[`toHaveBeenCalledTimes fails if function called less than expected times 1`] = `
"<dim>expect(<red>spy</><dim>).toHaveBeenCalledTimes(<green>2</><dim>)

Expected spy to have been called <green>two times</>, but it was called <red>one time</>."
`;

exports[`toHaveBeenCalledTimes fails if function called more than expected times 1`] = `
"<dim>expect(<red>spy</><dim>).toHaveBeenCalledTimes(<green>2</><dim>)

Expected spy to have been called <green>two times</>, but it was called <red>three times</>."
`;

exports[`toHaveBeenCalledTimes passes if function called equal to expected times 1`] = `
"<dim>expect(<red>spy</><dim>).not.toHaveBeenCalledTimes(<green>2</><dim>)

Expected spy not to be called <green>two times</>, but it was called exactly <red>two times</>."
`;

exports[`toHaveBeenCalledTimes verifies that actual is a Spy 1`] = `
"<dim>expect(<red>jest.fn()</><dim>)[.not].toHaveBeenCalledTimes(<dim>)

<red>jest.fn()</> value must be a mock function or spy.
Received:
  function: <red>[Function fn]</>"
`;

exports[`toHaveBeenCalledWith works only on spies or jest.fn 1`] = `
"<dim>expect(<red>jest.fn()</><dim>)[.not].toHaveBeenCalledWith(<dim>)

<red>jest.fn()</> value must be a mock function or spy.
Received:
  function: <red>[Function fn]</>"
`;

exports[`toHaveBeenCalledWith works with jasmine.createSpy and arguments that don't match 1`] = `
"<dim>expect(<red>spy</><dim>).toHaveBeenCalledWith(<green>expected</><dim>)

Expected spy to have been called with:
  <green>[\\"foo\\", \\"bar\\"]</>
But it was called with:
  <red>[\\"foo\\", \\"bar1\\"]</>"
`;

exports[`toHaveBeenCalledWith works with jasmine.createSpy and arguments that match 1`] = `
"<dim>expect(<red>spy</><dim>).not.toHaveBeenCalledWith(<green>expected</><dim>)

Expected spy not to have been called with:
  <green>[\\"foo\\", \\"bar\\"]</>"
`;

exports[`toHaveBeenCalledWith works with jasmine.createSpy and many arguments 1`] = `
"<dim>expect(<red>spy</><dim>).not.toHaveBeenCalledWith(<green>expected</><dim>)

Expected spy not to have been called with:
  <green>[\\"foo\\", \\"bar\\"]</>"
`;

exports[`toHaveBeenCalledWith works with jasmine.createSpy and many arguments that don't match 1`] = `
"<dim>expect(<red>spy</><dim>).toHaveBeenCalledWith(<green>expected</><dim>)

Expected spy to have been called with:
  <green>[\\"foo\\", \\"bar\\"]</>
But it was called with:
  <red>[\\"foo\\", \\"bar3\\"]</>, <red>[\\"foo\\", \\"bar2\\"]</>, <red>[\\"foo\\", \\"bar1\\"]</>"
`;

exports[`toHaveBeenCalledWith works with jasmine.createSpy when not called 1`] = `
"<dim>expect(<red>spy</><dim>).toHaveBeenCalledWith(<green>expected</><dim>)

Expected spy to have been called with:
  <green>[\\"foo\\", \\"bar\\"]</>
But it was <red>not called</>."
`;

exports[`toHaveBeenCalledWith works with jest.fn and arguments that don't match 1`] = `
"<dim>expect(<red>jest.fn()</><dim>).toHaveBeenCalledWith(<green>expected</><dim>)

Expected mock function to have been called with:
  <green>[\\"foo\\", \\"bar\\"]</>
But it was called with:
  <red>[\\"foo\\", \\"bar1\\"]</>"
`;

exports[`toHaveBeenCalledWith works with jest.fn and arguments that match 1`] = `
"<dim>expect(<red>jest.fn()</><dim>).not.toHaveBeenCalledWith(<green>expected</><dim>)

Expected mock function not to have been called with:
  <green>[\\"foo\\", \\"bar\\"]</>"
`;

exports[`toHaveBeenCalledWith works with jest.fn and many arguments 1`] = `
"<dim>expect(<red>jest.fn()</><dim>).not.toHaveBeenCalledWith(<green>expected</><dim>)

Expected mock function not to have been called with:
  <green>[\\"foo\\", \\"bar\\"]</>"
`;

exports[`toHaveBeenCalledWith works with jest.fn and many arguments that don't match 1`] = `
"<dim>expect(<red>jest.fn()</><dim>).toHaveBeenCalledWith(<green>expected</><dim>)

Expected mock function to have been called with:
  <green>[\\"foo\\", \\"bar\\"]</>
But it was called with:
  <red>[\\"foo\\", \\"bar3\\"]</>, <red>[\\"foo\\", \\"bar2\\"]</>, <red>[\\"foo\\", \\"bar1\\"]</>"
`;

exports[`toHaveBeenCalledWith works with jest.fn when not called 1`] = `
"<dim>expect(<red>jest.fn()</><dim>).toHaveBeenCalledWith(<green>expected</><dim>)

Expected mock function to have been called with:
  <green>[\\"foo\\", \\"bar\\"]</>
But it was <red>not called</>."
`;

exports[`toHaveBeenLastCalledWith works only on spies or jest.fn 1`] = `
"<dim>expect(<red>jest.fn()</><dim>)[.not].toHaveBeenLastCalledWith(<dim>)

<red>jest.fn()</> value must be a mock function or spy.
Received:
  function: <red>[Function fn]</>"
`;

exports[`toHaveBeenLastCalledWith works with jasmine.createSpy and arguments that don't match 1`] = `
"<dim>expect(<red>spy</><dim>).toHaveBeenLastCalledWith(<green>expected</><dim>)

Expected spy to have been last called with:
  <green>[\\"foo\\", \\"bar\\"]</>
But it was last called with:
  <red>[\\"foo\\", \\"bar1\\"]</>"
`;

exports[`toHaveBeenLastCalledWith works with jasmine.createSpy and arguments that match 1`] = `
"<dim>expect(<red>spy</><dim>).not.toHaveBeenLastCalledWith(<green>expected</><dim>)

Expected spy to not have been last called with:
  <green>[\\"foo\\", \\"bar\\"]</>"
`;

exports[`toHaveBeenLastCalledWith works with jasmine.createSpy and many arguments 1`] = `
"<dim>expect(<red>spy</><dim>).not.toHaveBeenLastCalledWith(<green>expected</><dim>)

Expected spy to not have been last called with:
  <green>[\\"foo\\", \\"bar\\"]</>"
`;

exports[`toHaveBeenLastCalledWith works with jasmine.createSpy and many arguments that don't match 1`] = `
"<dim>expect(<red>spy</><dim>).toHaveBeenLastCalledWith(<green>expected</><dim>)

Expected spy to have been last called with:
  <green>[\\"foo\\", \\"bar\\"]</>
But it was last called with:
  <red>[\\"foo\\", \\"bar3\\"]</>
and <red>two more calls</>."
`;

exports[`toHaveBeenLastCalledWith works with jasmine.createSpy when not called 1`] = `
"<dim>expect(<red>spy</><dim>).toHaveBeenLastCalledWith(<green>expected</><dim>)

Expected spy to have been last called with:
  <green>[\\"foo\\", \\"bar\\"]</>
But it was <red>not called</>."
`;

exports[`toHaveBeenLastCalledWith works with jest.fn and arguments that don't match 1`] = `
"<dim>expect(<red>jest.fn()</><dim>).toHaveBeenLastCalledWith(<green>expected</><dim>)

Expected mock function to have been last called with:
  <green>[\\"foo\\", \\"bar\\"]</>
But it was last called with:
  <red>[\\"foo\\", \\"bar1\\"]</>"
`;

exports[`toHaveBeenLastCalledWith works with jest.fn and arguments that match 1`] = `
"<dim>expect(<red>jest.fn()</><dim>).not.toHaveBeenLastCalledWith(<green>expected</><dim>)

Expected mock function to not have been last called with:
  <green>[\\"foo\\", \\"bar\\"]</>"
`;

exports[`toHaveBeenLastCalledWith works with jest.fn and many arguments 1`] = `
"<dim>expect(<red>jest.fn()</><dim>).not.toHaveBeenLastCalledWith(<green>expected</><dim>)

Expected mock function to not have been last called with:
  <green>[\\"foo\\", \\"bar\\"]</>"
`;

exports[`toHaveBeenLastCalledWith works with jest.fn and many arguments that don't match 1`] = `
"<dim>expect(<red>jest.fn()</><dim>).toHaveBeenLastCalledWith(<green>expected</><dim>)

Expected mock function to have been last called with:
  <green>[\\"foo\\", \\"bar\\"]</>
But it was last called with:
  <red>[\\"foo\\", \\"bar3\\"]</>
and <red>two more calls</>."
`;

exports[`toHaveBeenLastCalledWith works with jest.fn when not called 1`] = `
"<dim>expect(<red>jest.fn()</><dim>).toHaveBeenLastCalledWith(<green>expected</><dim>)

Expected mock function to have been last called with:
  <green>[\\"foo\\", \\"bar\\"]</>
But it was <red>not called</>."
`;
